### PR TITLE
bpo-36333: Fix leaks in pymain_init and _PyPreConfig_ReadFromArgv

### PR DESCRIPTION
**Pull request by matrixise**
https://bugs.python.org/issue36333

### CODE REVIEW

File: Python/pystate.c [hunk 1]
```diff
@@ -92,6 +92,11 @@ _PyRuntimeState_Fini(_PyRuntimeState *runtime)
         runtime->interpreters.mutex = NULL;
     }
 
+    if (runtime->xidregistry.mutex != NULL) {
+        PyThread_free_lock(runtime->xidregistry.mutex);
+        runtime->xidregistry.mutex = NULL;
```
Comment (vstinner) [inline Python/pystate.c:97]: Please write a separated PR to this change. I don't know well this code, so I'm not sure how xidregistry is supposed to be used.

File: Python/preconfig.c [hunk 2]
```diff
@@ -758,6 +758,7 @@ _PyPreConfig_ReadFromArgv(_PyPreConfig *config, const _PyArgv *args)
 done:
     if (init_ctype_locale != NULL) {
         setlocale(LC_CTYPE, init_ctype_locale);
+        PyMem_RawFree(init_ctype_locale);
```
Comment (vstinner) [inline Python/preconfig.c:761]: Oops, I missed that in the big refactoring I did last ones. Nicely spotted ;-)

File: Modules/main.c [hunk 3]
```diff
@@ -363,6 +363,7 @@ pymain_init(const _PyArgv *args, PyInterpreterState **interp_p)
 
     err = config_read_write(config, args, preconfig);
     if (_Py_INIT_FAILED(err)) {
+        _PyRuntimeState_Fini(&_PyRuntime);
```
Comment (vstinner) [inline Modules/main.c:366]: That's not the right place to cleanup things. I redesigned deeply Py_Main() to replace exit() call in mutliple layers of function calls, far from Py_Main(), with _PyInitError which allows to give back the control flow to Py_Main(). I did that to be able to cleanup Python properly.

**Review by vstinner — CHANGES_REQUESTED** [review]
Hum, I think that I will rewrite your PR because I'm in the middle of a big refactoring, and the correct fix for your bug requires more fixes.

But please write a separated PR for _PyRuntimeState_Fini.

**Comment by matrixise** [comment]
This PR is superseded by #12400. @vstinner thank your review.

File: Python/preconfig.c (hunk 2 above)
Comment (btharper) [reply Python/preconfig.c:761]: There's a similar leak in _PyPreConfig_Read, https://github.com/btharper/cpython/commit/81e617d18b1f64f24895f5c83fadfc655f05d760 I didn't see any more cases grepping for setlocale calls

**Pull request closed without merging**